cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -223,35 +230,65 @@ private void processNotification(AppMetadataStore appMetadataStore, Notification
-          return;
+          recorded = appMetadataStore.recordProgramResumed(programId, runId, messageIdBytes);
+          break;
         case COMPLETED:
         case KILLED:
           if (endTimeSecs == -1) {
             LOG.warn("Ignore program killed notification for program {} without end time specified, {}",
                      programRunId, notification);
             return;
           }
-          appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, null, messageIdBytes);
-          return;
+          recorded =
+            appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, null, messageIdBytes);
+          break;
         case FAILED:
           if (endTimeSecs == -1) {
             LOG.warn("Ignore program failed notification for program {} without end time specified, {}",
                      programRunId, notification);
             return;
           }
           BasicThrowable cause = decodeBasicThrowable(properties.get(ProgramOptionConstants.PROGRAM_ERROR));
-          appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, cause, messageIdBytes);
-          return;
+          recorded =
+            appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, cause, messageIdBytes);
+          break;
         default:
           // This should not happen
           LOG.error("Unsupported program status {} for program {}, {}", programRunStatus, programRunId, notification);
+          return;
+      }
+      if (recorded) {
+        // Notification with status RESUMING is actually recorded as RUNNING
+        publishRecordedStatus(programRunId, programRunStatus == ProgramRunStatus.RESUMING ?
+          ProgramRunStatus.RUNNING : programRunStatus);
+      }
+    }
+
+    private void publishRecordedStatus(ProgramRunId programRunId, ProgramRunStatus status) {
+      Notification programStatusNotification =
+        new Notification(Notification.Type.PROGRAM_STATUS,
+                         ImmutableMap.of(ProgramOptionConstants.PROGRAM_RUN_ID, GSON.toJson(programRunId),
+                                         ProgramOptionConstants.PROGRAM_STATUS, status.name()));
+
+      try {
+        messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), topic, GSON.toJson(programStatusNotification));
+      } catch (Exception e) {
+        if (e instanceof ServiceUnavailableException) {
+          SAMPLING_LOG.warn("Failed to contact service {}. Will retry in next run.",
+                            ((ServiceUnavailableException) e).getServiceName(), e);
+        } else {
+          SAMPLING_LOG.warn("Failed to publish status '{}' of program run '{}' to TMS. Will retry in next run.",
+                            status, programRunId, e);
+        }
+        // Simply propagate the exception and let the caller to retry
+        Throwables.propagate(e);
```

Comment:
So if there is any exception not related to ServiceUnavailable, we will get stuck and not able to make any progress?

Reply:
should I just log the exception without publishing if it's not `ServiceUnavailable`?

Reply:
Actually, since it is a transactional publish we are doing here, there will be no exception thrown by the `publish` call. The try-catch would never get triggered. If the TMS is not available during tx commit time, the tx would failed.

Can you open a JIRA to add support for auto-retry in publisher via the ClientMessagingService?

Reply:
so no need of propagating?

Reply:
no need to have try-catch at all.